README.md

```diff
@@ -1,39 +1,70 @@
 # setup-haskell
 
-<p align="left">
-  <a href="https://github.com/actions/setup-haskell"><img alt="GitHub Actions status" src="https://github.com/actions/setup-haskell/workflows/Main%20workflow/badge.svg"></a>
-</p>
+[![GitHub Actions status](https://github.com/actions/setup-haskell/workflows/Main%20workflow/badge.svg)](https://github.com/actions/setup-haskell)
 
-This action sets up a Haskell environment for use in actions by:
+This action installs a specified version of GHC and Cabal. It can also install [Stack](https://haskellstack.org).
 
-- optionally installing a version of ghc and cabal and adding to PATH. Note that this action only uses versions of ghc and cabal already installed in the cache. The action will fail if no matching versions are found.
+Currently only Linux comes with [pre-installed versions of GHC and Cabal](https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md). Those will be used whenever possible.
```

Comment:
I didn't realize that the images came with GHC 8 all the way up to 8.8. I think this should be okay, though we'll need to keep track of when we need to upgrade (since 8.10 just landed). Is there a way to force an image to use `ghcup`?

Reply:
> I just gave this a quick review. Again, really appreciate the contributions here @jared-w!
> 
> I think the first thing to tackle is to move any installing of haskell tools over to https://github.com/actions/virtual-environments. This is critical for build time performance. This action (`setup-haskell`) should simply put the right tools on the path. Does that make sense?

Makes total sense. I've opened [actions/virtual-environments#645](https://github.com/actions/virtual-environments/issues/645) to address this. I do know that the setup-node action will fall back and download a full version of node if the version specified doesn't exist, so there's precedent for said fallback behavior.

The intention of most of this code was always to become the fallback behavior and have the images come with GHC across all supported operating systems. Since this supports installing as a fallback, it also means github could drop about 8 of their current GHC installs.

Reply:
> I didn't realize that the images came with GHC 8 all the way up to 8.8. I think this should be okay, though we'll need to keep track of when we need to upgrade (since 8.10 just landed). Is there a way to force an image to use `ghcup`?

The action currently doesn't have any code for forcing the usage of `ghcup` over the PPA, but will fall back if GHC wasn't already installed and trying the PPA didn't work. I'm open to suggestions on how to tweak/change that.

As for the actual virtual environment for the runners, I'm not sure how the upgrade process currently works; I'm assuming you don't want to automatically download the "three latest major releases" when building the images and would rather specify them explicitly?